File: backend/services/AuthService.py
import time
from ..store import AuthStore
from ..api import oauth, users


class AuthService:
    def __init__(self):
        self.authData = AuthStore()

        self.authData.load()

    def tokenExpired(self):
        if self.authData.access_token != None and self.authData.expires_at != None:
            return time.time() > self.authData.expires_at

        return False

    async def tryRefreshAuthData(self):
        resp = await oauth.requestAccessTokenByRefreshToken(self.authData.refresh_token)

        if resp.status == 200:
            data = resp.json()

            self.authData.refresh(data)

            return True

        print("Auth data invalidated, cleaning!")

        self.authData.refresh(None)
        return False

    async def validateAuthData(self):
        if not self.authData.refresh_token == None:
            if self.tokenExpired():
                if not self.tryRefreshAuthData():
                    return False

            resp = await users.getCurrentUser(self.authData.access_token)

            if resp.status == 200:
                return resp.json()

            self.authData.refresh(None)
            return False

    def isLoggedIn(self):
        auth = self.authData

        return auth.access_token != None and auth.refresh_token != None

    async def login(self, code: str):
        resp = await oauth.requestAccessToken(code)

        if resp.status == 200:
            data = resp.json()
            print(data)
            self.authData.refresh(data)

            return True

        return False

    async def logout(self):
        if self.isLoggedIn():
            self.auth.refresh(null)
            return True

        return False

    async def getValidAccessToken(self):
        if not self.tokenExpired():
            return self.authData.access_token

        if not await self.tryRefreshAuthData():
            return None

        return self.authData.access_token

File: api/src/modules/auth/auth.service.ts
import { Inject, Injectable } from "@nestjs/common";
import AuthStore from "src/class/AuthStore";
import { OauthService } from "src/oauth/oauth.service";
import { UsersService } from "src/users/users.service";

@Injectable()
export class AuthService {
  private authData: AuthStore;

  constructor(
    @Inject(OauthService)
    private readonly oauthService: OauthService,
    @Inject(UsersService)
    private readonly usersService: UsersService
  ) {
    this.authData = new AuthStore();
    this.authData.load();
  }

  private tokenExpired(): boolean {
    this.authData.load();
    if (
      this.authData.access_token !== undefined &&
      this.authData.expires_at !== undefined
    ) {
      return Date.now() > this.authData.expires_at;
    }
    return false;
  }

  async tryRefreshAuthData(): Promise<boolean> {
    this.authData.load();
    const resp = await this.oauthService.requestAccessTokenByRefreshToken(
      this.authData.refresh_token
    );

    if (resp.status === 200) {
      const data = resp.data;
      this.authData.refresh(data);
      return true;
    }

    console.log("Auth data invalidated, cleaning!");
    this.authData.refresh(null);
    return false;
  }

  async validateAuthData(): Promise<any | false> {
    this.authData.load();
    if (this.authData.refresh_token) {
      if (this.tokenExpired() && !(await this.tryRefreshAuthData())) {
        return false;
      }
      const resp = await this.usersService.getCurrentUser(
        this.authData.access_token
      );

      if (resp.status === 200) {
        return resp.data;
      }

      this.authData.refresh(null);
      return false;
    }
  }

  isLoggedIn(): boolean {
    this.authData.load();
    const auth = this.authData;

    return auth.access_token !== undefined && auth.refresh_token !== undefined;
  }

  async login(code: string): Promise<boolean> {
    this.authData.load();
    const resp = await this.oauthService.requestAccessToken(code);

    if (resp.status === 200) {
      const data = resp.data;
      console.log(data);
      this.authData.refresh(data);
      return true;
    }

    return false;
  }

  async logout(): Promise<boolean> {
    this.authData.load();
    if (this.isLoggedIn()) {
      this.authData.refresh(null);
      return true;
    }

    return false;
  }

  async getValidAccessToken(): Promise<string | null> {
    this.authData.load();
    if (!this.tokenExpired()) {
      return this.authData.access_token;
    }

    if (!(await this.tryRefreshAuthData())) {
      return null;
    }

    return this.authData.access_token;
  }
}
